### DISCUSSION TITLE
Should winget manifests wait for installation to complete?

### DISCUSSION THREAD

**Discussion by othrayte** [microsoft/winget-cli#1456] [Q&A] [answered]
When trying to install VisualStudio (Microsoft.VisualStudio.2019.Community) via winget I discovered that winget invokes the installer without the option specified to wait for the installation to complete. Installing VisualStudio actually installs the VisualStudio installer and asks it to install VS. The VS installer has it's own GUI that opens to show progress of installing the product, but the VS installer is installed so that process returns and the winget install command finishes.

The problem is that you can ask winget to install VS on the command line but if you have no graphical interface or you invoked winget with the `--silent` option the command will return before the package you asked for is installed and you have no way of knowing when it completes. This is especially problematic for automated installations.

There is a work around by passing `--wait` to the installer via `--override`, however this forces the passing of all the options that winget would normally specify, negative much of the benefit of using winget. e.g. `winget install Microsoft.VisualStudio.2019.Community --override "--wait --quiet --productId Microsoft.VisualStudio.Product.Community --channelId VisualStudio.16.Release --channelUri https://aka.ms/vs/16/release/channel --add Microsoft.VisualStudio.Workload.Universal"`; note that even `--silent` need to be done using `--override`.

So, my question is should package manifests be setup to wait for the target package to be installed, in this case VS Community? i.e. should the VS community manifest specify `--wait` in the custom installer switches section.

**Comment by jedieaston** [comment, marked as answer]
We used to pass `—wait` but the installer would timeout during automated validation. 

I can put it back but some tuning might need to be done to the pipeline.

**Reply by othrayte** [reply]
Thanks for that insight.

So it sounds like packages in general should wait but there was a problem with waiting in this case. I don't find the issue very surprising, I've had trouble in the past getting the vs installer exes to work properly from the command line, which is why I was hoping to use winget as a simpler way to install vs and not have to worry about maintaining the set of options to pass to make it play nice.

Unfortunately the waiting was only my first problem with the VS community winget package as in my situation have to override the options anyway to prevent it installing the universal workload as I need just the compiler. Ideally I would just install the build tools version and wait there would be perfect however I'm using cmake downstream and for some unknown reason it refuses to consider vs build tools to contain a valid compiler (even if installing the same/equivalent workloads).

I think it would be good if winget could wait for the installation but as it might not help me at the moment anyway and I assume the timeout issue affects other people I'll leave it to you to decide it you want to invest the time in changing it.

**Reply by jedieaston** [reply]
Just wanted to update you on this: the issue with validating long running installers seems to have been resolved. I've opened PRs to add the `--wait` switch to all of the VS installers. Hopefully, by the end of today, you should be good to go.
